### PR TITLE
[None][refactor] Unify PrimTS paged decode entry points

### PR DESCRIPTION
**Pull request by yuxianq**
## 📌 Description

Unify PrimTS FMHA and MLA paged decode around the existing reusable wrappers:
`batch_decode_with_paged_kv_cache` and `batch_mla_decode_with_paged_kv_cache`
now support both internally allocated and caller-owned workspace through the
same wrapper `plan()` / `run()` path.

Current head: `1eb01e13a852dbb7dfe7cc10d761fd0563f46d6a`.
The branch was rebased onto `main` at `711d3bc94e6d2130af8607320eec612899600084`.
#5082 is now merged into `main`; its `@flashinfer_experimental_api` decorators
and experimental documentation/trace policy are preserved.

- When workspace is omitted, calculate its size, allocate it, and pass it to the
  wrapper. Caller-provided workspace uses the same execution path; there is no
  separate caller-workspace launch helper.
- Keep `validate=True` as the default. Trusted execution uses `validate=False`,
  caller-owned workspace, explicit `max_kv_len`, and an explicit packed-Q bound
  when applicable. For capture, also provide output storage and warm the exact
  topology outside capture. This avoids metadata readback and tensor allocation
  in the warmed, preallocated path.
- Keep FMHA convenience-call length specialization, fixed/packed Q, supported
  dtypes/cache layouts, scales, masks/windows, and output layout. FMHA planning
  preserves initialized external scratch with `initialize_workspace=False`;
  callers still own workspace initialization, capacity, and lifetime.
- Remove the superseded standalone decode entry points and their exports/lazy
  imports. Workspace-sizing helpers, prepared plans, and reusable wrappers remain.
- Separate device normalization from runtime CUDA validation and reuse the same
  resolver in both validated and trusted planning. Preserve explicit device
  validation in the affected block-sparse callers.
- Remove capture-state checks from the four internal policy/compiler cache
  functions. Warmup remains the caller's responsibility; public convenience
  API capture checks remain.
- Update existing attention tests, trace templates/examples, and documentation.
  No new benchmark file or dedicated experimental test file is included.
  No attention-kernel implementation or TensorRT-LLM source changes are included.

## 🔍 Related Issues

Follow-up to #4829. Experimental API annotation changes were merged separately
as #5082 and are now inherited from `main`.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` (using the existing installation).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files`.

Changed-file/staged pre-commit checks passed, including mypy and Ruff.
`git diff --check` passed. The full-repository hook run is not claimed.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Latest local validation for the code committed as `1eb01e13`, on an NVIDIA A40:

| Suite | Passed | Skipped |
| --- | ---: | ---: |
| FMHA decode, MLA decode, block-sparse attention, sparse metadata | 452 | 367 |
| Trace registry and template consistency | 697 | 0 |
| Total | **1,149** | **367** |

No failures in these local runs. Hardware-dependent tests were skipped;
the A40 does not validate the Blackwell attention kernels.

Attention targets (CUDA was initialized before invoking pytest for the
CUTLASS policy tests):

```python
import torch
import pytest

torch.cuda.init()
raise SystemExit(pytest.main([
    "-q",
    "tests/attention/test_attention_ts_decode.py",
    "tests/attention/test_attention_ts_mla_decode.py",
    "tests/attention/test_attention_ts_block_sparse.py",
    "tests/attention/test_attention_ts_q_token_kv_block_sparse_metadata.py",
    "--disable-warnings", "--tb=short",
]))
```

Trace targets:

```sh
CUDA_VISIBLE_DEVICES='' python -m pytest -q \
  tests/trace/test_template_registry.py \
  tests/trace/test_fi_trace_template_consistency.py
```

These runs used the existing development runtime with the worktree on
`PYTHONPATH` and `FLASHINFER_DISABLE_VERSION_CHECK=1`.

### B200 performance rerun — 2026-09-17

**Result: no meaningful GPU latency regression was observed in the eight sampled
explicit-workspace cases, but the allocating MLA convenience API has a
reproducible wall-time regression. This is not an unconditional no-regression
signoff.**

- Before: PR base `711d3bc94e6d2130af8607320eec612899600084`.
- After: current shared-wrapper implementation `1eb01e13a852dbb7dfe7cc10d761fd0563f46d6a`, clean worktree.
- NVIDIA B200, 1000 W power limit; driver 610.57.04; Python 3.12.3;
  PyTorch 2.12.0a0+5aff3928d8.nv26.05 / CUDA 13.2; CUTLASS DSL 4.7.0.
- Same GPU and process for each A/B pair. Baseline API modules were loaded from an
  immutable Git archive in an isolated module namespace; common runtime support
  is shared. Kernel Python source digests match exactly:
  `a31bc4e9fcc915aa7d08a8b4c7d90144fb4bceb09c795324234503f44bdfefc1`.
- Qwen-shaped FMHA: Hq=28, Hkv=4, D=128. DeepSeek-Lite-shaped MLA: H=32,
  latent=512, RoPE=64. Page size=32. BF16 uses fixed Q=1, B=1/64;
  FP8 E4M3 uses packed Q=4, B=4, with FP16 FMHA output / BF16 MLA output.
- Five alternating-order rounds, 50 eager warmup calls per API; GPU timing uses
  FlashInfer CUDA-graph timing, 10 warmup replays and 100 measured replays per
  round, 10 calls per graph, warm L2. Reported values are medians of the five
  round medians. One-time compilation and correctness checks are excluded;
  per-call planning, validation, and allocation remain in the eager/API timings.
- The old standalone APIs are compared with the new canonical APIs using
  caller-owned workspace, explicit KV/Q bounds, preallocated output, and
  `validate=False`. Each pair uses identical input **and identical
  workspace/output addresses**, sequentially, to remove allocation-placement
  bias. Host API time averages 100 calls per round and excludes the final
  synchronization; it is not a pure-Python self-time metric.
- All eight cases passed bitwise old/new eager-output equality, convenience-output
  equality, and explicit-API CUDA-graph replay equality. This checks equivalence,
  not an independent numerical oracle or the full GPU test suite.

#### Explicit-workspace API

| Case | KV / Q | GPU µs, before → after | Latency Δ | Kernel ktoken/s, before → after | Host API µs, before → after | Host Δ |
| --- | --- | ---: | ---: | ---: | ---: | ---: |
| Qwen B1 BF16 | 1536 / 1 | 9.776 → 9.776 | +0.00% | 102.29 → 102.29 | 98.27 → 70.49 | -28.27% |
| Qwen B64 BF16 | 1536 / 1 | 33.334 → 33.430 | +0.29% | 1919.94 → 1914.43 | 101.21 → 72.97 | -27.90% |
| DeepSeek-Lite B1 BF16 | 1536 / 1 | 9.981 → 9.981 | +0.00% | 100.19 → 100.19 | 214.77 → 196.92 | -8.31% |
| DeepSeek-Lite B64 BF16 | 1536 / 1 | 29.638 → 29.840 | +0.68% | 2159.36 → 2144.77 | 65.49 → 47.96 | -26.77% |
| Qwen B4 FP8 | 257 / 4 | 5.885 → 5.885 | +0.00% | 2718.87 → 2718.87 | 100.12 → 70.26 | -29.83% |
| Qwen B4 FP8 | 4097 / 4 | 13.667 → 13.770 | +0.75% | 1170.69 → 1161.98 | 100.63 → 70.60 | -29.84% |
| DeepSeek-Lite B4 FP8 | 257 / 4 | 14.282 → 14.282 | +0.00% | 1120.32 → 1120.32 | 73.85 → 56.06 | -24.09% |
| DeepSeek-Lite B4 FP8 | 4097 / 4 | 17.968 → 17.968 | +0.00% | 890.47 → 890.47 | 74.61 → 56.63 | -24.10% |

Throughput is `B × Q / GPU latency`, expressed in thousands of query tokens/s
for **one attention operation**, not end-to-end model-serving throughput.
GPU latency changes are 0.00% to +0.75%, with overlapping per-round ranges.
Explicit-workspace host API time decreases 8.3%–29.8%.

<details>
<summary>Per-round GPU median ranges (10 calls/graph)</summary>

| Case | KV | Before GPU µs, min–max | After GPU µs, min–max |
| --- | ---: | ---: | ---: |
| qwen B1 | 1536 | 9.571–9.981 | 9.571–9.776 |
| qwen B64 | 1536 | 33.022–33.533 | 33.018–33.533 |
| deepseek B1 | 1536 | 9.978–9.981 | 9.779–9.981 |
| deepseek B64 | 1536 | 29.027–30.096 | 28.822–30.250 |
| qwen B4 | 257 | 5.680–5.885 | 5.680–5.885 |
| qwen B4 | 4097 | 13.667–13.875 | 13.667–13.872 |
| deepseek B4 | 257 | 14.282–14.282 | 14.282–14.282 |
| deepseek B4 | 4097 | 17.968–17.971 | 17.968–17.973 |

</details>

#### Allocating convenience API — remaining regression

Here both revisions use their canonical API with workspace and KV bound omitted,
default validation enabled, and preallocated output. These calls perform
allocation/metadata synchronization, so they are measured eagerly, not captured.

| Case | KV | Before µs | After µs | Change |
| --- | ---: | ---: | ---: | ---: |
| Qwen B1 BF16 | 1536 | 226.65 | 223.43 | -1.42% |
| Qwen B64 BF16 | 1536 | 809.84 | 544.88 | -32.72% |
| DeepSeek-Lite B1 BF16 | 1536 | 314.80 | 353.51 | +12.30% |
| DeepSeek-Lite B64 BF16 | 1536 | 448.30 | 462.90 | +3.26% |
| Qwen B4 FP8 | 257 | 266.50 | 265.04 | -0.55% |
| Qwen B4 FP8 | 4097 | 349.40 | 309.05 | -11.55% |
| DeepSeek-Lite B4 FP8 | 257 | 197.22 | 204.81 | +3.85% |
| DeepSeek-Lite B4 FP8 | 4097 | 239.01 | 246.42 | +3.10% |

MLA convenience calls regress **3.1%–12.3%** in this run. The final synchronization
adds less than 0.4 µs/call to these medians, so the synchronized eager timings show
the same conclusion. Qwen convenience calls are not slower in the measured matrix.

#### Independent core repeat

A fresh process repeated all four BF16 cases with **100 calls/graph**, keeping
the same five-round / 100-replay / identical-storage controls. Absolute GPU times
depend on graph length; compare before/after within each row.

| BF16 case | Before GPU µs | After GPU µs | GPU Δ | Convenience host µs, before → after | Convenience Δ |
| --- | ---: | ---: | ---: | ---: | ---: |
| qwen B1 | 9.4051 | 9.4054 | +0.00% | 226.43 → 223.40 | -1.33% |
| qwen B64 | 32.9664 | 32.9570 | -0.03% | 814.84 → 543.18 | -33.34% |
| deepseek B1 | 9.3030 | 9.2928 | -0.11% | 316.21 → 350.64 | +10.89% |
| deepseek B64 | 31.6141 | 31.5218 | -0.29% | 447.73 → 465.18 | +3.90% |

The MLA convenience regression persists: +10.89% (B1), +3.90% (B64).
No source fix was made as part of this measurement request. The shared-wrapper
implementation therefore meets the sampled GPU-latency goal, but still needs
follow-up on allocating MLA API overhead before claiming no regression across
all call modes.

Artifacts are outside the PR/worktrees:
`/home/yuxianq/scratch/prims-ts-perf-results/20260917-pr5084-1eb01e13/`
contains `bench.py`, `run-gpu.sh`, immutable `baseline/`, and
`core.log`, `extended.log`, `confirm.log` (all round samples, eager timings,
and reusable-wrapper controls). No benchmark file was added back to the PR.

Reproduction in the recorded Blackwell runtime, with the current worktree on
`PYTHONPATH` and `FLASHINFER_DISABLE_VERSION_CHECK=1`:

```sh
python /path/to/artifacts/bench.py --baseline /path/to/artifacts/baseline
python /path/to/artifacts/bench.py --baseline /path/to/artifacts/baseline --extended
python /path/to/artifacts/bench.py --baseline /path/to/artifacts/baseline --graph-calls 100
```

### Remaining validation

- The B200 microbenchmark above validates sampled old/new output equivalence,
  CUDA-graph replay, and performance on the current head. The complete Blackwell
  correctness suite and end-to-end model-serving performance were not rerun.
- Allocating MLA convenience API overhead still regresses; see the measured
  results above. A blanket no-performance-regression signoff remains open.
- As of 2026-09-17, the latest [public GPU CI summary](https://github.com/flashinfer-ai/flashinfer/actions/runs/35203216623/job/105142716868)
  reports **CI skipped — pending authorization**; the GPU matrix did not run.
  This is not a reported kernel-test failure. An authorized CI trigger is still needed.

## 🔬 Experimental Track

- [x] This PR changes existing `@flashinfer_experimental_api` entries.
  Experimental status comes from merged #5082; this refactor adds no new backend.
- [ ] A new tracking issue with owner and graduation plan has been provided.
- [ ] Core changes are limited to a thin entry point.
- [ ] Tests live in `tests/experimental/` and were validated on the intended hardware.
- [x] No AOT registration or automatic experimental backend selection is added.
- [ ] Experimental-lane test scope is declared below.

Per review, regression coverage stays in the existing `tests/attention/` and
`tests/trace/` suites; the added experimental test module was removed.
The experimental scope parser only accepts existing targets under
`tests/experimental/`, so a valid declaration for that lane remains unresolved.
The deleted target is not advertised as runnable, and unrelated experimental
tests are not substituted. The existing targets above can be requested explicitly
through the CI bot.

```experimental-tests
# No applicable target under tests/experimental/ is included in this PR.
# Existing attention/trace targets are listed above; experimental-lane scope remains open.
```

## Reviewer Notes

- Review workspace ownership/initialization and shared wrapper routing, especially
  trusted capture calls and FMHA length specialization.
- Confirm use of the existing attention/trace suites for this refactor.
- Sampled B200 parity/capture checks passed. Full GPU-suite signoff and the
  allocating MLA API performance regression remain outstanding.

## Summary by CodeRabbit

- **New Features**
  - Paged decode APIs now support caller-provided workspace buffers, explicit maximum KV lengths, and optional validation.
  - Trusted execution paths can support CUDA graph capture when required inputs and workspace are supplied.
  - Decode planning can optionally skip validation and workspace initialization.

- **API Changes**
  - Legacy standalone PrimTS decode entry points and exports have been removed.
  - Existing paged-cache APIs now provide the consolidated decode interface.

- **Bug Fixes**
  - Improved runtime-device resolution and validation across decode and block-sparse operations.
  - Preserved published plans when replanning fails.

- **Documentation**
  - Updated API references and examples for the paged-cache interfaces.

### CODE REVIEW

**Review by PerkzZheng — APPROVED** [review]

**Comment by qsang-nv** [comment]
**Minor cleanup: two trace dispatchers and their dedicated factories/dicts have no production consumer after this PR.**

Their only `@flashinfer_experimental_api(trace=...)` consumers — `prims_ts_batch_decode_with_kv_cache` and `prims_ts_batch_mla_decode_with_kv_cache` — are deleted here, leaving two closed clusters in `flashinfer/trace/templates/attention.py`:

- `_make_prims_ts_decode_trace` (L1093) → `_PRIMS_TS_DECODE_TRACES` (L1211) → `prims_ts_decode_trace_dispatch` (L1225)
- `_make_prims_ts_decode_mla_trace` (L1572) → `_PRIMS_TS_DECODE_MLA_TRACES` (L1666) → `prims_ts_decode_mla_trace_dispatch` (L1675)

Each factory is used only by its own dict, each dict only by its own dispatcher, and the matching `fi_trace_out/` recordings are already deleted here.

Test impact: `test_template_signature_consistency` / `test_template_axes_covered` already excluded experimental APIs at the PR base (`tests/trace/template_registry.py` L138), so this removal does not reduce their coverage. `test_attention_ts_trace_constraints_match_cache_axes` checks template consistency but does not verify that a live API consumes the templates.

Please remove both orphaned dispatchers, their dedicated factories/template dictionaries, and the corresponding references in `test_attention_ts_trace_constraints_match_cache_axes`, while retaining checks for the live dispatchers.

**Comment by yuxianq** [comment]
@qsang-nv Good catch, I have removed these unused code, please review again. Thanks~

**Review by qsang-nv — APPROVED** [review]
